File: app/src/main/java/bassamalim/hidaya/models/MyLocation.java
package bassamalim.hidaya.models;

import android.location.Location;
import android.os.Build;

import java.io.Serializable;

public class MyLocation implements Serializable {

    private final float accuracy;
    private final long time;
    private final double altitude;
    private final float bearing;
    private float bearingAccuracyDegrees;
    private final long elapsedRealtimeNanos;
    private double elapsedRealtimeUncertaintyNanos;
    private final double latitude;
    private final double longitude;
    private final String provider;
    private final float speed;
    private float speedAccuracyMetersPerSecond;
    private float verticalAccuracyMeters;

    public MyLocation(Location loc) {
        accuracy = loc.getAccuracy();
        time = loc.getTime();
        altitude = loc.getAltitude();
        bearing = loc.getBearing();
        elapsedRealtimeNanos = loc.getElapsedRealtimeNanos();
        latitude = loc.getLatitude();
        longitude = loc.getLongitude();
        provider = loc.getProvider();
        speed = loc.getSpeed();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            bearingAccuracyDegrees = loc.getBearingAccuracyDegrees();
            speedAccuracyMetersPerSecond = loc.getSpeedAccuracyMetersPerSecond();
            verticalAccuracyMeters = loc.getVerticalAccuracyMeters();
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
                elapsedRealtimeUncertaintyNanos = loc.getElapsedRealtimeUncertaintyNanos();
        }
    }


    public float getAccuracy() {
        return accuracy;
    }

    public long getTime() {
        return time;
    }

    public double getAltitude() {
        return altitude;
    }

    public float getBearing() {
        return bearing;
    }

    public float getBearingAccuracyDegrees() {
        return bearingAccuracyDegrees;
    }

    public long getElapsedRealtimeNanos() {
        return elapsedRealtimeNanos;
    }

    public double getElapsedRealtimeUncertaintyNanos() {
        return elapsedRealtimeUncertaintyNanos;
    }

    public double getLatitude() {
        return latitude;
    }

    public double getLongitude() {
        return longitude;
    }

    public String getProvider() {
        return provider;
    }

    public float getSpeed() {
        return speed;
    }

    public float getSpeedAccuracyMetersPerSecond() {
        return speedAccuracyMetersPerSecond;
    }

    public float getVerticalAccuracyMeters() {
        return verticalAccuracyMeters;
    }

    public static Location toLocation(MyLocation myLoc) {
        Location loc = new Location("provider");

        loc.setAccuracy(myLoc.getAccuracy());
        loc.setTime(myLoc.getTime());
        loc.setAltitude(myLoc.getAltitude());
        loc.setBearing(myLoc.getBearing());
        loc.setElapsedRealtimeNanos(myLoc.getElapsedRealtimeNanos());
        loc.setLatitude(myLoc.getLatitude());
        loc.setLongitude(myLoc.getLongitude());
        loc.setProvider(myLoc.getProvider());
        loc.setSpeed(myLoc.getSpeed());
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            loc.setBearingAccuracyDegrees(myLoc.getBearingAccuracyDegrees());
            loc.setSpeedAccuracyMetersPerSecond(myLoc.getSpeedAccuracyMetersPerSecond());
            loc.setVerticalAccuracyMeters(myLoc.getVerticalAccuracyMeters());
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
                loc.setElapsedRealtimeUncertaintyNanos(loc.getElapsedRealtimeUncertaintyNanos());
        }

        return loc;
    }

}

File: app/src/main/java/bassamalim/hidaya/models/MyLocation.kt
package bassamalim.hidaya.models

import android.location.Location
import android.os.Build
import java.io.Serializable

data class MyLocation(val loc: Location) : Serializable {

    val accuracy = loc.accuracy
    val time = loc.time
    val altitude = loc.altitude
    val bearing = loc.bearing
    var bearingAccuracyDegrees = 0f
    val elapsedRealtimeNanos = loc.elapsedRealtimeNanos
    private var elapsedRealtimeUncertaintyNanos = 0.0
    val latitude = loc.latitude
    val longitude = loc.longitude
    val provider: String? = loc.provider
    val speed = loc.speed
    var speedAccuracyMetersPerSecond = 0f
    var verticalAccuracyMeters = 0f

    init {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            bearingAccuracyDegrees = loc.bearingAccuracyDegrees
            speedAccuracyMetersPerSecond = loc.speedAccuracyMetersPerSecond
            verticalAccuracyMeters = loc.verticalAccuracyMeters
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) elapsedRealtimeUncertaintyNanos =
                loc.elapsedRealtimeUncertaintyNanos
        }
    }

    companion object {    // companion object is like a container for static methods
        fun toLocation(myLoc: MyLocation): Location {
            val loc = Location("provider")
            loc.accuracy = myLoc.accuracy
            loc.time = myLoc.time
            loc.altitude = myLoc.altitude
            loc.bearing = myLoc.bearing
            loc.elapsedRealtimeNanos = myLoc.elapsedRealtimeNanos
            loc.latitude = myLoc.latitude
            loc.longitude = myLoc.longitude
            loc.provider = myLoc.provider
            loc.speed = myLoc.speed
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                loc.bearingAccuracyDegrees = myLoc.bearingAccuracyDegrees
                loc.speedAccuracyMetersPerSecond = myLoc.speedAccuracyMetersPerSecond
                loc.verticalAccuracyMeters = myLoc.verticalAccuracyMeters
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
                    loc.elapsedRealtimeUncertaintyNanos = loc.elapsedRealtimeUncertaintyNanos
            }
            return loc
        }
    }
}